test/examples/ucp_client_server.c

```diff
@@ -0,0 +1,436 @@
+                        ucp_listener_h *listener)
+{
+    struct sockaddr_in listen_addr;
+    ucs_status_t status;
+
+    context->ep = NULL;
+
+    set_listen_addr(&listen_addr);
+    status = server_listen(ucp_worker, (const struct sockaddr*)&listen_addr,
+                           sizeof(listen_addr), listener,
+                           context);
+
+    if (status == UCS_OK) {
+        while (context->ep == NULL) {
+            ucp_worker_progress(ucp_worker);
+        }
+    }
+
+    return status;
+}
+
+/**
+ * Initialize the client side. Create an endpoint from the client side to be
+ * connected to the remote server (to the given IP).
+ */
+static ucs_status_t start_client(ucp_worker_h ucp_worker, const char *ip,
+                                 ucp_ep_h *client_ep)
+{
+    ucp_ep_params_t ep_params;
+    struct sockaddr_in connect_addr;
+    ucs_status_t status;
+
+    set_connect_addr(ip, &connect_addr);
+
+    ep_params.field_mask       = UCP_EP_PARAM_FIELD_FLAGS     |
+                                 UCP_EP_PARAM_FIELD_SOCK_ADDR |
+                                 UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE;
+    ep_params.err_mode         = UCP_ERR_HANDLING_MODE_PEER;
+    ep_params.flags            = UCP_EP_PARAMS_FLAGS_CLIENT_SERVER;
+    ep_params.sockaddr.addr    = (struct sockaddr*)&connect_addr;
+    ep_params.sockaddr.addrlen = sizeof(connect_addr);
+
+    status = ucp_ep_create(ucp_worker, &ep_params, client_ep);
+    if (status != UCS_OK) {
+        fprintf(stderr, "failed to connect to %s\n", ip);
+    }
+
+    return status;
+}
+
+/**
+ * Verify the received message on the server side and print the result.
+ */
+static int verify_result(int *is_server, char *recv_message)
+{
+    int ret;
+
+    if (*is_server) {
+        if (!strcmp(recv_message, test_message)) {
+            printf("\n\n----- UCP TEST SUCCESS -------\n\n");
+            printf("%s", recv_message);
+            printf("\n\n------------------------------\n\n");
+
+            ret = 0;
+        } else {
+            printf("\n\n----- UCP TEST FAILURE -------\n\n");
+            printf("Client sent message:     %s\nServer received message: %s\n",
+                   test_message, recv_message);
+            printf("\n\n------------------------------\n\n");
+
+            ret = -1;
+        }
+    } else {
+        printf("\n\n-----------------------------------------\n\n");
+        printf("Client sent message: \n%s.\nlength: %ld\n",
+               test_message, TEST_STRING_LEN);
+        printf("\n-----------------------------------------\n\n");
+        ret = 0;
+    }
+
+    return ret;
+}
+
+/**
+ * Send and receive a message using the Stream API.
+ * The client sends a message to the server and waits until the send it completed.
+ * The server receives a message from the client and waits for its completion.
+ */
+static int send_recv_stream(ucp_worker_h ucp_worker, ucp_ep_h ep, int *is_server)
```

Comment:
seems like `is_server` can be passed by value

Reply:
@evgeny-leksikov yes, it definitely can. this is a forgotten leftover from a previous code, will update this of course, 10x.